### ISSUE TITLE
multi instance problem on titan when  ./xmlchange MULTI_DRIVER=TRUE

### ISSUE THREAD

**Issue by kevans32** [ESMCI/cime#2295]
I am trying out this multi instance capability in titan. Very cool!
I can run when ./xmlchange MULTI_DRIVER=FALSE but when 
 ./xmlchange MULTI_DRIVER=TRUE it seems to not know how to use the second set of tasks 

run command : aprun -S 8 -n 1280 -N 16 -d 1 /ccs/home/4ue/acme_scratch/cli115/FC5multi2/bld/acme.exe    >> acme.log.$LID 2>&1

But in output it does this:
(seq_comm_setcomm)  initialize ID (  1 GLOBAL          ) pelist   =     0   639     1 ( npes =   640) ( nthreads =  1)( suffix =)
Rank 526 [Wed Feb 14 12:04:40 2018] [c4-1c2s5n2] Fatal error in PMPI_Group_range_incl: Invalid argument, error stack:
PMPI_Group_range_incl(195)........: MPI_Group_range_incl(group=0x88000000, n=1, ranges=0x5b495c0, new_group=0x7fffffff3d44) failed
MPIR_Group_check_valid_ranges(323): **The 0th element of a range array ends at 1279 but must be nonnegative and less than 640**

**Here are my bash commands for a simple test run with 2 instances**
casename=FC5multi2
casedirectory=$PROJWORK/cli115/4ue                                             
machine=titan
projectname=cli115
compset=FC5
grid=ne30_ne30

module load python/2.7.9
code_dir=$HOME
code_tag=ACME
cd ${code_dir}/${code_tag}/cime/scripts

$code_dir/$code_tag/cime/scripts/create_newcase -case $casedirectory/$casename -mach $machine -project $projectname -compset $compset -res $grid

cd $casedirectory/$casename

  ./xmlchange NINST_ATM=2
  ./xmlchange NINST_LND=2
  ./xmlchange NINST_ICE=2
  ./xmlchange NINST_ROF=2
  ./xmlchange NINST_GLC=2
  ./xmlchange NINST_OCN=2
  ./xmlchange NINST_WAV=2
  ./xmlchange NINST_ESP=2

  ./xmlchange MULTI_DRIVER=TRUE (or FALSE)

  ./case.setup

cat <<EOF >> user_nl_cam_0001
 zmconv_c0_lnd = 0.007
EOF
cat <<EOF >> user_nl_cam_0002
 zmconv_c0_lnd = 0.008
EOF

  ./case.build
  ./case.submit

**Comment by jedwards4b** [comment]
Hi Kate,
In your create_newcase command try adding --ninst 2 --multi-driver 
let me know if that changes the result.

**Comment by kevans32** [comment]
Hey Jim! I did a fresh build using those commands as part of create_newcase instead and got the same result. My script is here on titan if you want to play:
/ccs/home/4ue/ACME/run_multiFC5.sh

**Comment by jedwards4b** [comment]
Sorry, I can't help you on titan.   I did test this on cheyenne and it works as expected.

**Comment by mvertens** [comment]
Hi Kate,

Can you try running your script on cheyenne to see if it works there.

Mariana

**Comment by kevans32** [comment]
ooh can I have access to Cheyenne?

**Comment by kevans32** [comment]
Is there any problem with me just using the FALSE option and asking for the right # cores for now on titan? I need to stick with titan because we will be using the GPU

**Comment by mvertens** [comment]
I think this should be fine. @jedwards4b can confirm.

**Comment by jedwards4b** [comment]
Lets see if we can find the problem- after your create_newcase (with multi-driver and ninst 2)
what is the result of ./pelayout ?  and ./preview_run?

**Comment by kevans32** [comment]
Comp  NTASKS  NTHRDS  ROOTPE
CPL :   1280/     1;      0
ATM :   1280/     1;      0
LND :   1280/     1;      0
ICE :   1280/     1;      0
OCN :   1280/     1;      0
ROF :   1280/     1;      0
GLC :   1280/     1;      0
WAV :   1280/     1;      0
ESP :      1/     1;      0
BATCH SUBMIT: (nodes 88)
    case.run -> qsub    -q batch -l walltime=02:00:00 -A cli115 .case.run 

MPIRUN: aprun -S 8 -n 1280 -N 16 -d 1 /ccs/home/4ue/acme_scratch/cli115/FC5multi2/bld/acme.exe    >> acme.log.$LID 2>&1

**Comment by jedwards4b** [comment]
Okay - I think that this is a problem with the particular way that titan batch job is set up 
It should be using 160 nodes, not 80.

**Comment by jedwards4b** [comment]
If you do ./xmlquery TOTALPES you should get 2560?

**Comment by kevans32** [comment]
TOTALPES: 1280... so nope

**Comment by jedwards4b** [comment]
@jgfouca is going to look at this when he gets a chance, you should be able to run without the --multi-driver option if you need to get going.

**Comment by kevans32** [comment]
thanks. I am good I think - still need to look at the output and test etc. If you need a simple test of this functionality let me know I should have one as a result of this...

**Comment by kevans32** [comment]
@jgfouca hmmm I might need this feature after all. I am creating an ensemble of serial jobs (of SCAM) so I am using a mpi-serial build using the --ninst=2 feature just like above with
When I do this, even with ./xmlchange MULTI_DRIVER=FALSE
I get a similar but not ID error:
MPI_Group_range_incl: more than 1 proc in group
which I think might be because E3SM thinks I am running in parallel since it sees I have allocated 2 processors and not one to EACH run to for multiple runs? I am wondering if I set the 
MULTI_DRIVER=TRUE to use the multiple instance feature,
then it will execute 2 independent jobs each with 1 proc successfully rather than trying to divide up 2 procs to 2 jobs, and in so doing, not correctly assigned all tasks in the case setup to 1 proc each. Does that logic makes sense? If not, should I open up a separate issue for the MPI error itself? THANKS

**Comment by jedwards4b** [comment]
@kevans32  - I don't think you can do that - if you are using multiple instances you need mpi to be able to communicate between them even if each one only uses a single task.

**Comment by kevans32** [comment]
Oh, so does this capability require MPI for tasks to talk to each other under the hood in the coupler or something? and If so what info is being exchanged? Can I turn off MPI for that part and have it happen sequentially?

**Comment by jedwards4b** [comment]
Anytime more than one task is needed MPI is used to coordinate the tasks, there really isn't any information being shared other than the comms that allow different instances to run on different tasks.   You can run scam with a single mpi task, you don't need to use mpi-serial.

**Comment by kevans32** [comment]
So can we not have this multi-instance be used without MPI and have each task run independently without MPI? The SCAM docs on confluence say it is supported with mpi-serial. If I turn on MPI, I get an 'invalid communicator'error. Do you know how to run SCAM with MPI on?

**Comment by jedwards4b** [comment]
You should be able to run scam with MPI - make sure that NTASKS=1 for all components.

**Comment by jedwards4b** [comment]
It runs with mpiserial but not with mpi and ntasks=1?   That doesn't make
any sense to me.

**Comment by gold2718** [comment]
Yes. I tried a default run of FSCAM with T42_T42 and it ran fine (nag on Hobart).
Then I tried:
```
./create_newcase --case ${case} --compset FSCAM --res T42_T42 --machine hobart --compiler nag --run-unsupported
cd ${case}
./xmlchange MPILIB=openmpi,NTASKS=1 
./case.setup && ./case.build && ./case.submit
```
and I get:
```
Fatal error in PMPI_Comm_size: Invalid communicator, error stack:
PMPI_Comm_size(112): MPI_Comm_size(comm=0x0, size=0x7ffe344b3f34) failed
PMPI_Comm_size(71).: Invalid communicator
```

**Comment by jedwards4b** [comment]
This is a bug we should fix, there is no logical reason it should occur.

**Comment by jgfouca** [comment]
CAM people need to fix SCAM with MPI+ntasks=1. Assigning to @singhbalwinder for lack of better options, feel free to pass along to the appropriate person.

**Comment by jedwards4b** [comment]
I think that this has been addressed and should be closed.

**Comment by jedwards4b** [comment]
closing as inactive
